src/frequenz/sdk/timeseries/formula_engine/_formula_engine.py

```diff
@@ -1020,6 +1021,52 @@ def min(
         """
         return self._push("min", other)
 
+    def consumption(
+        self,
+    ) -> (
+        HigherOrderFormulaBuilder[QuantityT]
+        | HigherOrderFormulaBuilder3Phase[QuantityT]
+    ):
+        """Apply the Consumption Operator.
+
+        The consumption operator returns either the identity if the power value is
+        positive or 0.
+
+        Returns:
+            A formula builder that can take further expressions, or can be built
+                into a formula engine.
+        """
+        self._steps.appendleft((TokenType.OPER, "("))
+        self._steps.append((TokenType.OPER, ")"))
+        self._steps.append((TokenType.OPER, "consumption"))
+        assert isinstance(
+            self, (HigherOrderFormulaBuilder, HigherOrderFormulaBuilder3Phase)
+        )
+        return self
+
+    def production(
+        self,
+    ) -> (
+        HigherOrderFormulaBuilder[QuantityT]
+        | HigherOrderFormulaBuilder3Phase[QuantityT]
+    ):
```

Comment:
Because these are unary operators, and our conversion to postfix can only deal with binary operators, these steps should call build from within, so that no further steps can be added on this layer.

Or an alternative is to turn these into binary operators by adding a second `0` operand that gets discarded by the corresponding formula steps.

Reply:
Why do you think it can only deal with binary operators? I don't think this is true and I've added tests that show that the combination works without flaws.

After all in the case of a binary operator the Formula Step simply pops one value from the formula stack and pushes the result back. Since we are adding brackets before adding the consumption or production operators we make sure that the formula, that we operate on, is evaluated first and then we apply these operators on the remaining value.

Reply:
true, that appears to be the case.  As long as the input to the infix-to-postfix algorithm looks like this: `((a + b) production) + c`, etc., it seems to work.  I think the brackets after production/consumption are also important.